## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [16]:
random_array= np.random.randint(0, 10, 10)
print(random_array)
random_array[(random_array>3)&(random_array<8)]*=-1
print(random_array)


[1 7 1 7 1 7 3 3 6 9]
[ 1 -7  1 -7  1 -7  3  3 -6  9]


**2.** Заменить максимальный элемент случайного массива на 0

In [22]:
a=np.random.randint(10, size=10)
print(a)
a[np.argmax(a)]=0
print(a)

[6 1 6 5 5 8 4 0 7 6]
[6 1 6 5 5 0 4 0 7 6]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход?? подается двумерный массив

In [4]:
a = np.array([[1, 2, 3], [4, 5, 6]])
b = np.array([[7, 8, 9], [10, 11, 12]])

grid_a, grid_b = np.meshgrid(a, b)
print(grid_a*grid_b)

[[ 7 14 21 28 35 42]
 [ 8 16 24 32 40 48]
 [ 9 18 27 36 45 54]
 [10 20 30 40 50 60]
 [11 22 33 44 55 66]
 [12 24 36 48 60 72]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [7]:
a=np.array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11],
       [12, 13, 14],
       [15, 16, 17],
       [18, 19, 20],
       [21, 22, 23]])

b=np.array([7,23,8,22]).reshape(2,2)

is_in_B = np.isin(a, np.unique(b))
a[np.any(is_in_B, axis=1)]


array([[ 6,  7,  8],
       [21, 22, 23]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
a=np.array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11],
       [ 3, 3, 3],
       [15, 16, 17],
       [18, 19, 20],
       [ 4, 4, 4],
       [24, 25, 26],
       [27, 28, 29]])

a[a.max(axis=1) !=a.min(axis=1)]

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11],
       [15, 16, 17],
       [18, 19, 20],
       [24, 25, 26],
       [27, 28, 29]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [8]:
a=np.array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11],
       [12, 13, 14],
       [15, 16, 17],
       [ 3,  4,  5],
       [18, 19, 20],
       [ 6,  7,  8],
       [12, 13, 14],
       [21, 22, 23]])

np.unique(a, axis=0)

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11],
       [12, 13, 14],
       [15, 16, 17],
       [18, 19, 20],
       [21, 22, 23]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [19]:
# without NymPy
x = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
n=1
asix0=len(x)
asix1=len(x[0])
diag=[x[i][i] for i in range(min(asix0,asix1)) if x[i][i]!=0]
for i in diag:
    n*=i
print(n)  

# with NymPy
x = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
diag=np.diagonal(x)
np.prod(diag[np.flatnonzero(diag)])

3


3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [157]:
# without NymPy
a = [1, 3, 2, 4]
b = [4, 1, 3, 2]
a, b = set(a), set(b)
print(a == b)

# with NymPy
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
np.array_equal(np.unique(x), np.unique(y))



* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [10]:
# without NymPy
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
if x[-1]==0:
    x=x[:-1]
elements_befor_zero=[x[i+1] for i in range(len(x)) if x[i]==0] 
max(elements_befor_zero)

# with NymPy
x = np.array([6, 2, 0, 3, 0, 0, 5, 7,0])
if x[-1]==0:
    x=x[:-1]
max(x[np.where(x==0)[0]+1])


5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [23]:
# without NymPy
x = [2, 2, 2, 3, 3, 3, 5]
count_elem=[x.count(i) for i in set(x)]
print(f'{list(set(x))}, {count_elem}')


# with NymPy
x = np.array([2, 2, 2, 3, 3, 3, 5]) 
np.unique(x, return_counts=True) 

[2, 3, 5], [3, 3, 1]


(array([2, 3, 5]), array([3, 3, 1], dtype=int64))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.


In [21]:
from scipy.spatial.distance import cdist
import time

a = np.array([[1, 2, 3], [4, 5, 6]])
b = np.array([[7, 8, 9], [10, 11, 12]])

# without NymPy
start_time_1 = time.time()
euclidean_distances_scipy = cdist(a, b, 'euclidean')
end_time_1 = time.time()
print(euclidean_distances_scipy)
print(f"Время работы (scipy.spatial.distance.cdist): {end_time_1 - start_time_1} сек")

# with NymPy
start_time_2 = time.time()
euclidean_distances_nympy=np.linalg.norm(a[:, np.newaxis, :] - b, axis=2)
end_time_2 = time.time()
print(euclidean_distances_nympy)
print(f"Время работы (NumPy): {end_time_1 - start_time_1} сек")



[[10.39230485 15.58845727]
 [ 5.19615242 10.39230485]]
Время работы (scipy.spatial.distance.cdist): 0.0009944438934326172 сек
[[10.39230485 15.58845727]
 [ 5.19615242 10.39230485]]
Время работы (NumPy): 0.0009944438934326172 сек


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [23]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [24]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [35]:
average_calories=calorie_stats.mean()
print(average_calories)
print(f'В среднем для различных марок хлопьев в одной порции содержится на {int(average_calories-60)} калорий больше,чем в порции CrunchieMunchies')

106.88311688311688
В среднем для различных марок хлопьев в одной порции содержится на 46 калорий больше,чем в порции CrunchieMunchies


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [26]:
calorie_stats_sorted=np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [27]:
median_calories=np.median(calorie_stats)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [32]:
for nth_percentile in range(100,0,-1):
    value = np.percentile(calorie_stats, nth_percentile)
    if value>60:
        print(f"{nth_percentile} перцентиль: {value} калорий")
    else:
        break

100 перцентиль: 160.0 калорий
99 перцентиль: 152.39999999999995 калорий
98 перцентиль: 150.0 калорий
97 перцентиль: 147.2 калорий
96 перцентиль: 140.0 калорий
95 перцентиль: 140.0 калорий
94 перцентиль: 140.0 калорий
93 перцентиль: 136.80000000000007 калорий
92 перцентиль: 130.0 калорий
91 перцентиль: 130.0 калорий
90 перцентиль: 124.00000000000006 калорий
89 перцентиль: 120.0 калорий
88 перцентиль: 120.0 калорий
87 перцентиль: 120.0 калорий
86 перцентиль: 120.0 калорий
85 перцентиль: 120.0 калорий
84 перцентиль: 120.0 калорий
83 перцентиль: 120.0 калорий
82 перцентиль: 120.0 калорий
81 перцентиль: 120.0 калорий
80 перцентиль: 120.0 калорий
79 перцентиль: 120.0 калорий
78 перцентиль: 120.0 калорий
77 перцентиль: 115.20000000000003 калорий
76 перцентиль: 110.0 калорий
75 перцентиль: 110.0 калорий
74 перцентиль: 110.0 калорий
73 перцентиль: 110.0 калорий
72 перцентиль: 110.0 калорий
71 перцентиль: 110.0 калорий
70 перцентиль: 110.0 калорий
69 перцентиль: 110.0 калорий
68 перцентиль: 110.

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [33]:
more_calories=np.mean(calorie_stats>60)*100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [34]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

с использованием различных методик анализа были выдвинуты следующие преимущества нашей марки, перед конкурентами: В среднем для различных марок хлопьев в одной порции содержится на 46 калорий больше, чем в порции CrunchieMunchies. Однако, этот результат не так впечатлит покупателя. 
Для лучшего развития маркетига компании лучше заявлять не о преимуществе перед среднем значением показателей конкурентов, а показать количество компаний, которых наша компания «обскакала» в данном показателе. Так, например, выгоднее сказать «CrunchieMunchies — ваши вкусные и умные закуски с калорийностью ниже, чем у 96% конкурентов! Придайте своему стилю нотку заботы о здоровье с CrunchieMunchies — лучший выбор для тех, кто ценит вкус и стремится к балансу между вкусом и питательностью». И уже чувствуется  победа над конкурентами !)
